sql: support GENERATED ... AS IDENTITY syntax under CREATE TABLE

ZhouXing19 opened the pull request:
This commit reorganizes the original `sql.processSerialInColumnDef`
method in `pkg/sql/serial.go` to build a serial for a column
that is specified with `GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY
(seq_option)` syntax 
under `CREATE TABLE` statement. The `seq_option`
is consistent with the sequence option syntax in `CREATE SEQUENCE`,
and we also support `CACHE` in `seq_option` for better performance.

Release note (sql change): support `GENERATED {ALWAYS | BY DEFAULT}
AS IDENTITY (seq_option)` 
syntax under CREATE TABLE

An `IDENTITY` column is an auto-incremented column based on an
automatically created sequence, with the same performance considerations
as the `CREATE SEQUENCE` syntax
(https://www.cockroachlabs.com/docs/stable/create-sequence.html#considerations).
The `seq_option` is consistent with the sequence option syntax
in `CREATE SEQUENCE`,and we also support `CACHE` in `seq_option`
for better performance
(https://www.cockroachlabs.com/docs/stable/create-sequence.html#cache-sequence-values-in-memory).

Hence, such a column can only be of integer type,
and is implicitly `NOT NULL`.
It is essentially the same as what `SERIAL` with
`serial_normalization=sql_sequence`,
except for user's access to override it.

A `GENERATED ALWAYS AS IDENTITY` column cannot be
overridden without specifying `OVERRIDING SYSTEM VALUE` in an
`INSERT`/`UPSERT`/`UPDATE` statement. This overriding issue cannot
be resolved by the `ON CONFLICT` syntax. Such a column can only be updated to
`DEFAULT`.

A `GENERATED BY DEFAULT AS IDENTITY` column allows being overridden
without specifying any extra syntax, and users is allowed to add
repeated values to such a column. It can also be updated to customized
expression, but only accepts integer type expression result.

`GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY` is also supported
under `ALTER TABLE ... ADD COLUMN ...` syntax.

This matches the PostgreSQL syntax:
https://www.postgresql.org/docs/current/sql-createtable.html

Release justification: although this supports a new feature,
the changes are high-benefit and have been under scrutiny for many weeks now,
so after discussing with other leads we feel fine about merging this during
stability period.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69107)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
Great job figuring out all these moving parts!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-:-MhPx7v-E1TA0kidlXlC:b-djte0v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/opt/cat/column.go, line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPu6gl-MMRXxcZsfJb:-MhPu6gl-MMRXxcZsfJc:bfc64ym) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L186)):*
> ```Go
> // GeneratedAsIdentityType reflects how the creation of this column
> // is associated with the GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY
> // token.
> ```

[nit] "token" doesn't seem like a good choice here, maybe just "syntax"
___
*[pkg/sql/opt/cat/column.go, line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPtofc0OV_R7j55lp8:-MhPtofc0OV_R7j55lp9:bcq4htb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L200)):*
> ```Go
> 	// GeneratedByDefaultAsIdentity column is created with
> 	// the GENERATED BY DEFAULT AS IDENTITY token.
> 	GeneratedByDefaultAsIdentity
> ```

This would be a good place to document what this actually means, especially the difference between Default and Always
___
*[pkg/sql/opt/cat/column.go, line 212 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPu45S385_GpBiGb2p:-MhPu45S385_GpBiGb2q:b-3wa2pq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L212)):*
> ```Go
> }
> 
> // GetGeneratedAsIdentityType returns how
> ```

[nit] wrapping
___
*[pkg/sql/opt/cat/column.go, line 215 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPtXPR2wC3WF0HjI_4:-MhPtXPR2wC3WF0HjI_5:b66y65w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L215)):*
> ```Go
> // the creation of this column is associated with
> // the GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY token.
> func (c *Column) GetGeneratedAsIdentityType() GeneratedAsIdentityType {
> ```

[nit] remove `Get`
___
*[pkg/sql/opt/cat/column.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPuJTzD-MGtiNVYnLO:-MhPuJTzD-MGtiNVYnLP:bxn1r8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L231)):*
> ```Go
> 	defaultExpr *string,
> 	computedExpr *string,
> 	generatedAsIDType GeneratedAsIdentityType,
> ```

[nit] I'd spell out `Identity`, since `ID` normally stands for "identifier"
___
*[pkg/sql/opt/cat/column.go, line 253 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPujpc5Reu68SwxZVi:-MhPujpc5Reu68SwxZVj:b-q5q2tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L253)):*
> ```Go
> 	}
> 	if defaultExpr != nil {
> 		c.defaultExpr = *defaultExpr
> ```

Should we assert here that generatedAsIdentityType is none? (same below)
___
*[pkg/sql/opt/cat/utils.go, line 360 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPtAY68LnVHt8sGYQy:-MhPtAY79YPSgsaZPEop:b-qi4ztf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/utils.go#L360)):*
> ```Go
> 	}
> 
> 	generatedAsIdentityType := col.GetGeneratedAsIdentityType()
> ```

[nit] I'd move this up so the stuff with brackets is at the end
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1328 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPvC3pABI9Obxub4wl:-MhPvC3r-1JihO-Z0GEi:bf62fbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ad3384d20f7d61fbf41b316f8d3ce96b92d9e9/pkg/sql/opt/optbuilder/mutation_builder.go#L1328)):*
> ```Go
> }
> 
> // checkColumnIsGeneratedAlwaysAsIdentity verifies that if current column
> ```

[nit] I can't quite parse this sentence, needs rephrasing. The name is also very confusing, shouldn't it be the opposite? (`checkColumnIsNotGeneratedAlwaysAsIdentity()`?)
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPwWys83_2fCJ7PzQk:-MhPwWys83_2fCJ7PzQl:b-4pkvra) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ad3384d20f7d61fbf41b316f8d3ce96b92d9e9/pkg/sql/opt/optbuilder/mutation_builder.go#L1349)):*
> ```Go
> }
> 
> // checkColumnIsGeneratedAlways verifies that if current column
> ```

[nit] consider renaming to something like "checkUpdateExpression()" and document that it throws an error if the expression is not DEFAULT and the column is GENERATED ALWAYS AS IDENTITY.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1352 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPw4mT6lHcOAMEH5_-:-MhPw4mU8BQMDHUAV35z:b-ok0ty6) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ad3384d20f7d61fbf41b316f8d3ce96b92d9e9/pkg/sql/opt/optbuilder/mutation_builder.go#L1352)):*
> ```Go
> // was created with the `GENERATED ALWAYS` token, which restricts
> // this column to only be updated to DEFAULT in the update statement.
> func checkColumnIsGeneratedAlways(col *cat.Column, updateExprs tree.UpdateExprs) {
> ```

Why are we passing all the update expressions here? Why are we looking at all of them?
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1354 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPvlJIEe2APSJk0a9u:-MhPvlJIEe2APSJk0a9v:bqyovie) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ad3384d20f7d61fbf41b316f8d3ce96b92d9e9/pkg/sql/opt/optbuilder/mutation_builder.go#L1354)):*
> ```Go
> func checkColumnIsGeneratedAlways(col *cat.Column, updateExprs tree.UpdateExprs) {
> 	// Check if this column is created with GENERATED ALWAYS token.
> 	if col.IsGeneratedAlwaysAsIdentity() || col.IsComputed() {
> ```

Why `IsComputed()`? Do we ever allow inserting computed values explicitly?
___
*[pkg/sql/opt/optbuilder/testdata/update, line 1694 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPwtpM9c24IFBplls5:-MhPwtpM9c24IFBplls6:b-32q4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ad3384d20f7d61fbf41b316f8d3ce96b92d9e9/pkg/sql/opt/optbuilder/testdata/update#L1694)):*
> ```
> 
> build
> UPDATE generated_as_identity SET c=DEFAULT;
> ```

We need tests with multiple update expressions.
___
*[pkg/sql/opt/testutils/testcat/create_table.go, line 128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPuDAfBtGW9XaxjZo5:-MhPuDAfBtGW9XaxjZo6:bwe1nqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/testutils/testcat/create_table.go#L128)):*
> ```Go
> 			&uniqueRowIDString,         /* defaultExpr */
> 			nil,                        /* computedExpr */
> 			cat.NotGeneratedAsIdentity, /* generatedAsIDType */
> ```

[nit] you don't need the `/* */` comment when the value is self-documenting.
___
*[pkg/sql/opt/testutils/testcat/create_table.go, line 604 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPuWLb0aIlSQZab8EC:-MhPuWLb0aIlSQZab8ED:b-2y659k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/testutils/testcat/create_table.go#L604)):*
> ```Go
> 	}
> 
> 	var generatedAsIDType cat.GeneratedAsIdentityType
> ```

[nit] can just do `generatedAsIdentityType := cat.NotGeneratedAsIdentity` and remove the `else`
___
*[pkg/sql/opt/testutils/testcat/create_table.go, line 611 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPud0345YlVtdyEx5X:-MhPud0345YlVtdyEx5Y:b-inrevq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/testutils/testcat/create_table.go#L611)):*
> ```Go
> 		case tree.GeneratedByDefault:
> 			generatedAsIDType = cat.GeneratedByDefaultAsIdentity
> 		}
> ```

default: panic?


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
___
*[pkg/sql/opt/cat/column.go, line 253 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPujpc5Reu68SwxZVi:-MhRKLMt1JIFzerfajSp:b-aolgc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L253)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Should we assert here that generatedAsIdentityType is none? (same below)
</blockquote></details>

Maybe not -- the `GENERATED ... AS IDENTITY` column has an implicit default sequence expression, so inherently `defaultExpr` and `IDENTITY` column are not exclusive. Also, the parser [already prevents](https://github.com/cockroachdb/cockroach/pull/69107/files#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67R537-R562) the case with both explicit default expression and `GENERATED ... AS IDENTITY` syntax, so I guess it's not necessary to add such an assertion here.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69107)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/opt/cat/column.go, line 253 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPujpc5Reu68SwxZVi:-MhRXScNBdjXz55XaYBR:bar5wk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L253)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Maybe not -- the `GENERATED ... AS IDENTITY` column has an implicit default sequence expression, so inherently `defaultExpr` and `IDENTITY` column are not exclusive. Also, the parser [already prevents](https://github.com/cockroachdb/cockroach/pull/69107/files#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67R537-R562) the case with both explicit default expression and `GENERATED ... AS IDENTITY` syntax, so I guess it's not necessary to add such an assertion here.
</blockquote></details>

Maybe just for the computedExpr case below then?

Assertions are not necessary by definition, but they help prevent bugs (especially in future changes).


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
I happened to be playing around with [Facebook's "Ent" ORM](https://entgo.io/) today because it was mentioned a few times in http://www.vldb.org/pvldb/vol14/p3014-cheng.pdf and because it's probably the most promising ORM in Go due to its embrace of code generation. The testing quickly ran into issues with the Postgres dialect, because by default Ent uses `GENERATED BY DEFAULT AS IDENTITY` for the primary key of every table. These PKs can be switched to a UUID, but that's a bit cumbersome. I remembered that I saw this PR fly by so I checked it out and, like magic, the entire thing came to life! It turns out that this is the last blocker for ~full Ent support, as demonstrated by https://github.com/nvanbenschoten/ent/commit/59f88f7eee40f7adb998ad08258b847a426b0bac.

That's super exciting. Keep up the great work @ZhouXing19!

ZhouXing19 commented:
___
*[pkg/sql/opt/cat/column.go, line 253 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPujpc5Reu68SwxZVi:-MhTmHX0EjZdvOx6o9oC:b-21ik1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L253)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Maybe just for the computedExpr case below then?

Assertions are not necessary by definition, but they help prevent bugs (especially in future changes).
</blockquote></details>

@raduberinde yeah makes sense, will do the computedExpr assertion. Thanks for pointing it out!


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
___
*[pkg/sql/opt/cat/column.go, line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPu6gl-MMRXxcZsfJb:-MhUBM3m3RrrgLM6Pkk7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L186)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] "token" doesn't seem like a good choice here, maybe just "syntax"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69107)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/opt/cat/column.go, line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPtofc0OV_R7j55lp8:-MhUBoUC0McHV48tGPBR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L200)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This would be a good place to document what this actually means, especially the difference between Default and Always
</blockquote></details>

Done.
___
*[pkg/sql/opt/cat/column.go, line 215 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPtXPR2wC3WF0HjI_4:-MhUD4ErBrZNc0gs5Q6g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L215)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] remove `Get`
</blockquote></details>

Done.
___
*[pkg/sql/opt/cat/column.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPuJTzD-MGtiNVYnLO:-MhUDBrr2rngOad4GX1X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L231)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] I'd spell out `Identity`, since `ID` normally stands for "identifier"
</blockquote></details>

Done.
___
*[pkg/sql/opt/cat/column.go, line 253 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPujpc5Reu68SwxZVi:-MhUCH3PBV6L4SkSDGCX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/column.go#L253)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

@raduberinde yeah makes sense, will do the computedExpr assertion. Thanks for pointing it out!
</blockquote></details>

Done.
___
*[pkg/sql/opt/cat/utils.go, line 360 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPtAY68LnVHt8sGYQy:-MhUDHukC3gMee5W3u7i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/cat/utils.go#L360)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] I'd move this up so the stuff with brackets is at the end
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1328 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPvC3pABI9Obxub4wl:-MhUD_9U8yyZRfHMvK5s:bgf24me) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ad3384d20f7d61fbf41b316f8d3ce96b92d9e9/pkg/sql/opt/optbuilder/mutation_builder.go#L1328)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] I can't quite parse this sentence, needs rephrasing. The name is also very confusing, shouldn't it be the opposite? (`checkColumnIsNotGeneratedAlwaysAsIdentity()`?)
</blockquote></details>

Rephrased the description here and rename the function. Thanks for pointing it out!
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPwWys83_2fCJ7PzQk:-MhUDpSR3T7c1KSa-Pop:bdgddp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ad3384d20f7d61fbf41b316f8d3ce96b92d9e9/pkg/sql/opt/optbuilder/mutation_builder.go#L1349)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] consider renaming to something like "checkUpdateExpression()" and document that it throws an error if the expression is not DEFAULT and the column is GENERATED ALWAYS AS IDENTITY.
</blockquote></details>

Good idea, thanks! Done.
___
*[pkg/sql/opt/testutils/testcat/create_table.go, line 128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPuDAfBtGW9XaxjZo5:-MhUE5hj1Kwt9kJKzLtb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/testutils/testcat/create_table.go#L128)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] you don't need the `/* */` comment when the value is self-documenting.
</blockquote></details>

Done.
___
*[pkg/sql/opt/testutils/testcat/create_table.go, line 604 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPuWLb0aIlSQZab8EC:-MhUE8my3Qvp7ICspSGi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/testutils/testcat/create_table.go#L604)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] can just do `generatedAsIdentityType := cat.NotGeneratedAsIdentity` and remove the `else`
</blockquote></details>

Done.
___
*[pkg/sql/opt/testutils/testcat/create_table.go, line 611 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPud0345YlVtdyEx5X:-MhUEBRi-9UFOuHpPLEH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade845f57f88abe97765095855aac4ee8234f15d/pkg/sql/opt/testutils/testcat/create_table.go#L611)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

default: panic?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
___
*[pkg/sql/opt/optbuilder/testdata/update, line 1694 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPwtpM9c24IFBplls5:-MhUWrIW9pcyI2XoXw1K:b-cbfshc) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ad3384d20f7d61fbf41b316f8d3ce96b92d9e9/pkg/sql/opt/optbuilder/testdata/update#L1694)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We need tests with multiple update expressions.
</blockquote></details>

Just added tests such as `UPDATE generated_as_identity SET a=10, b=(1+1);`. Are these what you meant by `multiple update expression`?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Thanks for the updates! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Would be good to wait for another review from the queries team (it's a big change and I probably missed some things).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-:-MhUZRWR3Qby_nEK2dp0:b-zakoqs)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/opt/cat/column.go, line 190 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhUXzIu6srGNmt6a6kX:-MhUXzIu6srGNmt6a6kY:bx54qdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc66e770720e124a371b5870f56a578840131489/pkg/sql/opt/cat/column.go#L190)):*
> ```Go
> // syntax.
> // The GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY generates an IDENTITY
> // column, which is based on an underlying sequence, and is auto-incremental
> ```

[nit] incremented
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1328 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhUYB6N6JVTwyXVMGNC:-MhUYB6N6JVTwyXVMGND:becyh20) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc66e770720e124a371b5870f56a578840131489/pkg/sql/opt/optbuilder/mutation_builder.go#L1328)):*
> ```Go
> }
> 
> // checkColumnIsNotGeneratedAlwaysAsIdentity verifies that if current column
> ```

[nit] verifies that the current column is not created as an IDENTITY colum
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1357 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhUYK1r9tYQsmVM2SMV:-MhUYK1r9tYQsmVM2SMW:b-9erisy) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc66e770720e124a371b5870f56a578840131489/pkg/sql/opt/optbuilder/mutation_builder.go#L1357)):*
> ```Go
> 		return
> 	default:
> 		if col.IsGeneratedAlwaysAsIdentity() {
> ```

[super nit] It would follow the high-level logic better if we first did `if col.IsGeneratedAlwaysAsIdentity()` and then checked the type
___
*[pkg/sql/opt/optbuilder/testdata/update, line 1694 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhPwtpM9c24IFBplls5:-MhUYbvxAD86xxHgjkrE:b-k614gu) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ad3384d20f7d61fbf41b316f8d3ce96b92d9e9/pkg/sql/opt/optbuilder/testdata/update#L1694)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Just added tests such as `UPDATE generated_as_identity SET a=10, b=(1+1);`. Are these what you meant by `multiple update expression`?
</blockquote></details>

Yes, thanks!


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69107)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @rafiss, and @RichardJCai)

___
*[pkg/sql/opt/cat/column.go, line 190 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhUXzIu6srGNmt6a6kX:-MhV0Go5AvXrnwtAnW5Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc66e770720e124a371b5870f56a578840131489/pkg/sql/opt/cat/column.go#L190)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] incremented
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1328 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhUYB6N6JVTwyXVMGNC:-MhV0IG01_NBrWicpUgu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc66e770720e124a371b5870f56a578840131489/pkg/sql/opt/optbuilder/mutation_builder.go#L1328)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] verifies that the current column is not created as an IDENTITY colum
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1357 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhUYK1r9tYQsmVM2SMV:-MhV0JJaEavaqPPQeFlY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc66e770720e124a371b5870f56a578840131489/pkg/sql/opt/optbuilder/mutation_builder.go#L1357)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[super nit] It would follow the high-level logic better if we first did `if col.IsGeneratedAlwaysAsIdentity()` and then checked the type
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
this looks great! i just had small nits

can you update the pg_catalog and information_schema as well?
- here https://github.com/cockroachdb/cockroach/blob/65457f9df618177bfd56096c9be0194db2ef5cf8/pkg/sql/pg_catalog.go#L389
- and here https://github.com/cockroachdb/cockroach/blob/65457f9df618177bfd56096c9be0194db2ef5cf8/pkg/sql/information_schema.go#L502
- feel free to modify those and add a logic test in this commit or a separate one. whichever you prefer

finally, could you add more details in the release note? (also the category should be `sql change` instead of `SQL`).

- mention that ALTER table also supports generated as identity too
- describe the difference between BY DEFAULT vs ALWAYS
- mention that it automatically creates a sequence, and that the same performance considerations as sequences apply (https://www.cockroachlabs.com/docs/stable/create-sequence.html#considerations)
- also mention that it supports sequence options and that CACHE can be used for better performance: https://www.cockroachlabs.com/docs/stable/create-sequence.html#cache-sequence-values-in-memory


```
Release note (sql change): support GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY syntax under
CREATE TABLE
```


> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-:-MhVByLr44mH_cwvhtbY:bjyy65t)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/alter_column_type.go, line 96 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhV8EST1xVefd2KLN0h:-MhV8EST1xVefd2KLN0i:b3it430) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/alter_column_type.go#L96)):*
> ```Go
> 
> 	// Special handling for IDENTITY column to make sure it cannot be altered into
> 	// a non-interger type.
> ```

nit: "non-integer"
___
*[pkg/sql/serial.go, line 201 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhV981_EFtKwF-VKyfU:-MhV981_EFtKwF-VKyfV:b-fv62ui) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/serial.go#L201)):*
> ```Go
> // for a column under CREATE TABLE.
> // This is a helper method for processSerialLikeInColumnDef.
> func (p *planner) processGenAsIdInColumnDef(
> ```

nit: rename this to `processGeneratedAsIdentityColumnDef`
___
*[pkg/sql/serial.go, line 214 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhV9Z85BtxT-xWUgX0J:-MhV9Z85BtxT-xWUgX0K:b-7mug7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/serial.go#L214)):*
> ```Go
> 	// TODO(janexing): discuss to generate a regular serial we should use
> 	// sessiondata.SerialUsesSQLSequences or sessiondata.SerialUsesCachedSQLSequences.
> 	curSerialNormalizationMode := sessiondatapb.SerialUsesSQLSequences
> ```

i think using this mode is the right approach, no need to keep the TODO

also, since we support `sequence_options` for GeneratedAsIdentity columns, users can add sequence caching themselves if they want
___
*[pkg/sql/logictest/testdata/logic_test/alter_table, line 1932 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhVAWad6A_pDdDJEqVr:-MhVAWaeEBpwtwBBxYOz:bnrz6eh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/logictest/testdata/logic_test/alter_table#L1932)):*
> ```
> 
> statement ok
> ALTER TABLE t ADD COLUMN b INT GENERATED BY DEFAULT AS IDENTITY (START 1 INCREMENT 3);
> ```

could you also add a test for the `CACHE` option here? it can be pretty simple, basically just make sure the CACHE option is parsed

i haven't looked into it, but you could look at `TestCachedSequences` in sequence_test.go and see if you can test a CACHEd GeneratedAsIdentity column


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> most of my comments are in the two previous PRs. Nice work!!

> Reviewed 1 of 3 files at r1, 42 of 42 files at r11, 18 of 18 files at r12, 15 of 15 files at r13, 18 of 18 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-:-MhVep7I6ZcdqIPuU72d:b-pwri11)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RichardJCai, and @ZhouXing19)

___
*[pkg/sql/catalog/tabledesc/column.go, line 197 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhVd6FVA_wHllt9B7ma:-MhVd6FWD2nNxHKb1vKF:bx2fohp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/catalog/tabledesc/column.go#L197)):*
> ```Go
> 
> // IsGeneratedAlwaysAsIdentity returns true iff the columns is created
> // with GENERATED ALWAYS AS IDENTITY token.
> ```

nit: why did you change this to token? To me, token implies a single word
___
*[pkg/sql/logictest/testdata/logic_test/create_table, line 538 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhVdoIl-pOllPTV-_Sb:-MhVdoIl-pOllPTV-_Sc:b-3nzsp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/logictest/testdata/logic_test/create_table#L538)):*
> ```
> gen_always_as_id_seqopt  CREATE TABLE public.gen_always_as_id_seqopt (
>                          a INT8 NULL,
>                          b INT8 NOT NULL GENERATED ALWAYS AS IDENTITY,
> ```

nit: add a TODO to print out sequence (if you don't want to make the change in this PR)
___
*[pkg/sql/parser/sql.y, line 6772 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhVeStF05RTsAdU99CD:-MhVeStF05RTsAdU99CE:b-vqgz40) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/parser/sql.y#L6772)):*
> ```Yacc
>   }
> | generated_by_default_as IDENTITY '(' opt_sequence_option_list ')'
> 	{
> ```

nit: remove tabs


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69107)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @rafiss, @RichardJCai, and @rytaft)

___
*[pkg/sql/alter_column_type.go, line 96 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhV8EST1xVefd2KLN0h:-MhnrcEpCVaN1feqdyeh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/alter_column_type.go#L96)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: "non-integer"
</blockquote></details>

Done.
___
*[pkg/sql/serial.go, line 201 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhV981_EFtKwF-VKyfU:-MhnrehY0pp5LLAfoOEW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/serial.go#L201)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: rename this to `processGeneratedAsIdentityColumnDef`
</blockquote></details>

Done.
___
*[pkg/sql/serial.go, line 214 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhV9Z85BtxT-xWUgX0J:-Mhnrg--DRUOQ4W506yZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/serial.go#L214)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think using this mode is the right approach, no need to keep the TODO

also, since we support `sequence_options` for GeneratedAsIdentity columns, users can add sequence caching themselves if they want
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/column.go, line 197 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhVd6FVA_wHllt9B7ma:-MhnriQA5SF9TyT_g7tE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/catalog/tabledesc/column.go#L197)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: why did you change this to token? To me, token implies a single word
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/alter_table, line 1932 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhVAWad6A_pDdDJEqVr:-MhnrjuFCDkYUMgxT23W:b-s8b68) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/logictest/testdata/logic_test/alter_table#L1932)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could you also add a test for the `CACHE` option here? it can be pretty simple, basically just make sure the CACHE option is parsed

i haven't looked into it, but you could look at `TestCachedSequences` in sequence_test.go and see if you can test a CACHEd GeneratedAsIdentity column

</blockquote></details>

Done, tests are added [here](https://github.com/cockroachdb/cockroach/pull/69107/commits/64cb71f8faa40dde3e0ad8adf0052547f79bcf6f#diff-fa3f37d689d1cc46fa50188c8b4b808e7f2b467e4424d45c47c72e57be8ff976).
___
*[pkg/sql/logictest/testdata/logic_test/create_table, line 538 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhVdoIl-pOllPTV-_Sb:-MhnrsVWEhDwEJwkeoQV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/logictest/testdata/logic_test/create_table#L538)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: add a TODO to print out sequence (if you don't want to make the change in this PR)
</blockquote></details>

Done.
___
*[pkg/sql/parser/sql.y, line 6772 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhVeStF05RTsAdU99CD:-Mhns1-U57oi42M5p-oU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d4984dcbf4591fb83959b71f621a3d9bd612e34/pkg/sql/parser/sql.y#L6772)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: remove tabs
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented on pkg/sql/logictest/testdata/logic_test/pg_catalog:
```diff
@@ -351,6 +351,8 @@ CREATE TABLE constraint_db.t1 (
   k VARCHAR(10) COLLATE sv,
   l INT AS (a * b) VIRTUAL,
+  m INT GENERATED ALWAYS AS IDENTITY,
+  n INT GENERATED BY DEFAULT AS IDENTITY,
```
@rafiss I added these two columns in `constraint_db.t1`, but ` gmake testlogic TESTCONFIG=fakedist-spec-planning FILES=pg_catalog TESTFLAGS=-rewrite` does not rewrite the file.

rafiss replied:
hm, what if you try without that `TESTCONFIG` variable? (also try with `FILES='pg_catalog|information_schema'` since that file might also need an update)

rytaft reviewed (approved):
Do you need to add some validation for the new sequence option? Otherwise <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 5 of 22 files at r16, 5 of 15 files at r17, 38 of 40 files at r18, 5 of 15 files at r21, 40 of 40 files at r22, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-:-Mho5Bvx11h8rS6DfcPN:b-oq7yrl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[-- commits, line 52 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho-XKO52g_GXwTaTo4:-Mho-XKO52g_GXwTaTo5:b10iau6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/-- commits#L52)):*
Looks like you need to fill this in
___
*[-- commits, line 139 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho-_PeCYsvDJcrTkM2:-Mho-_Pf7eOevQJlr5bo:b-89a4l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/-- commits#L139)):*
Ditto
___
*[pkg/clusterversion/cockroach_versions.go, line 305 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho-fiWECq_z7jlEl1E:-Mho-fiWECq_z7jlEl1F:ba0v9m8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/pkg/clusterversion/cockroach_versions.go#L305)):*
> ```Go
> 	UseKeyEncodeForHashShardedIndexes
> 	// GeneratedAsIdentity is the syntax support for `GENERATED {ALWAYS | BY
> 	//DEFAULT} AS IDENTITY` under `CREATE TABLE` syntax.
> ```

nit: add a space between // and GENERATED
___
*[pkg/sql/information_schema.go, line 459 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho01bc8Xx0t2_s6Wnf:-Mho01bc8Xx0t2_s6Wng:b-enjtfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/pkg/sql/information_schema.go#L459)):*
> ```Go
> 							"generated by default as identity")
> 					} else {
> 						panic(fmt.Errorf(
> ```

nit: don't panic, but return `errors.AssertionFailedf` here
___
*[pkg/sql/pg_catalog.go, line 409 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho0Svp-GWlDW51O8--:-Mho0Svp-GWlDW51O8-0:bbmnfav) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/pkg/sql/pg_catalog.go#L409)):*
> ```Go
> 					generatedAsIdentityType = "b"
> 				} else {
> 					panic(fmt.Errorf(
> ```

nit: don't panic, instead return errors.AssertionFailedf here
___
*[pkg/sql/opt/optbuilder/testdata/create_table, line 92 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho4AgG20XxftfV4kGV:-Mho4AgG20XxftfV4kGW:b-rc9f9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/pkg/sql/opt/optbuilder/testdata/create_table#L92)):*
> ```
> ----
> create-table
>  └── CREATE TABLE generated_as_identity_with_seq_opt (a INT8 UNIQUE, b INT8 NOT NULL GENERATED ALWAYS AS IDENTITY ( START 2 INCREMENT 3 CACHE 10 ) , c INT8 NOT NULL GENERATED BY DEFAULT AS IDENTITY ( START 2 INCREMENT 3 CACHE 10 ) )
> ```

nit: is there a way you can remove the spaces so it looks like: `(START 2 INCREMENT 3 CACHE 10)`
___
*[pkg/sql/parser/sql.y, line 6794 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho3stg1zZyyv9goI1L:-Mho3stg1zZyyv9goI1M:b-610j03) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/pkg/sql/parser/sql.y#L6794)):*
> ```Yacc
>   {
> 	   $$.val = &tree.GeneratedByDefAsIdentity{
> 	     SeqOptions: $4.seqOpts(),
> ```

nit: still a couple tabs


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/information_schema.go:
```diff
@@ -498,16 +513,15 @@ https://www.postgresql.org/docs/9.5/infoschema-columns.html`,
-					colComputed,                       // generation_expression
+					yesOrNoDatum(column.IsGeneratedAsIdentity()), // is_identity
+					colGeneratedAsIdentity,                       // identity_generation
+					tree.DNull,                                   // identity_start
```
can we fill out these other columns? the start, increment, max, min, and cycle should all be available on the sequence options

rafiss replied:
hm, i see that it looks like we'd need to re-parse the `column.GetGeneratedAsIdentitySequenceOption()` string. let's not worry about it for right now

ZhouXing19 committed:
sql: support GENERATED ... AS IDENTITY syntax under CREATE TABLE

This commit reorganizes the original `sql.processSerialInColumnDef`
method in `pkg/sql/serial.go` to build a serial for a column
that is specified with `GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY
(seq_option)` syntax under `CREATE TABLE` statement. The `seq_option`
is consistent with the sequence option syntax in `CREATE SEQUENCE`,
and we also support `CACHE` in `seq_option` for better performance.

Release note (sql change): support `GENERATED {ALWAYS | BY DEFAULT}
AS IDENTITY (seq_option)` syntax under CREATE TABLE

An `IDENTITY` column is an auto-incremented column based on an
automatically created sequence, with the same performance considerations
as the `CREATE SEQUENCE` syntax
(https://www.cockroachlabs.com/docs/stable/create-sequence.html#considerations).
The `seq_option` is consistent with the sequence option syntax
in `CREATE SEQUENCE`,and we also support `CACHE` in `seq_option`
for better performance
(https://www.cockroachlabs.com/docs/stable/create-sequence.html#cache-sequence-values-in-memory).

Hence, such a column can only be of integer type,
and is implicitly `NOT NULL`.
It is essentially the same as what `SERIAL` with
`serial_normalization=sql_sequence`,
except for user's access to override it.

A `GENERATED ALWAYS AS IDENTITY` column cannot be
overridden without specifying `OVERRIDING SYSTEM VALUE` in an
`INSERT`/`UPSERT`/`UPDATE` statement. This overriding issue cannot
be resolved by the `ON CONFLICT` syntax. Such a column can only be updated to
`DEFAULT`.

A `GENERATED BY DEFAULT AS IDENTITY` column allows being overridden
without specifying any extra syntax, and users is allowed to add
repeated values to such a column. It can also be updated to customized
expression, but only accepts integer type expression result.

`GENERATED {ALWAYS | BY DEFAULT} AS IDENTITY` is also supported
under `ALTER TABLE ... ADD COLUMN ...` syntax.

This matches the PostgreSQL syntax:
https://www.postgresql.org/docs/current/sql-createtable.html

Release justification: although this supports a new feature,
the changes are high-benefit and have been under scrutiny for many weeks now,
so after discussing with other leads we feel fine about merging this during
stability period.

ZhouXing19 reviewed (commented):
@rytaft Do you mean by [logictest](https://github.com/cockroachdb/cockroach/pull/69107/commits/fe03dce0c294d01d34cba029af5c0902bbf5fc18#diff-9f2dd101471409a892e6774c1e8a876db2d72f8d47d9e467513d9ae0ccc27f88R1961-R1994) with sequence options? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-:-Mhq7MVJ3qU_gA3_9xUM:b-sr9ugy)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @rafiss, @RichardJCai, and @rytaft)

___
*[-- commits, line 52 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho-XKO52g_GXwTaTo4:-Mhq6etU8aJjhq8Y8BGT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/-- commits#L52)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Looks like you need to fill this in
</blockquote></details>

Done.
___
*[-- commits, line 139 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho-_PeCYsvDJcrTkM2:-Mhq6fWD9trB6fEv63PK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/-- commits#L139)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Done.
___
*[pkg/clusterversion/cockroach_versions.go, line 305 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho-fiWECq_z7jlEl1E:-Mhq6irH4cJICoM3q7s4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/pkg/clusterversion/cockroach_versions.go#L305)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: add a space between // and GENERATED
</blockquote></details>

Done.
___
*[pkg/sql/information_schema.go, line 459 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho01bc8Xx0t2_s6Wnf:-Mhq6s-x3NLQ4vDtD9_J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/pkg/sql/information_schema.go#L459)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: don't panic, but return `errors.AssertionFailedf` here
</blockquote></details>

Done.
___
*[pkg/sql/information_schema.go, line 518 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhnmWUt6S8D28yi2DJn-r22-518:-Mhq6t386Q6Eo7gfukRG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2ed3facd7f6374e63b15fb2b243dec849b303d3/pkg/sql/information_schema.go#L518)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm, i see that it looks like we'd need to re-parse the `column.GetGeneratedAsIdentitySequenceOption()` string. let's not worry about it for right now, but adding a TODO would be useful
</blockquote></details>

Done.
___
*[pkg/sql/pg_catalog.go, line 409 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho0Svp-GWlDW51O8--:-Mhq6vjSCx1fhpeXARTq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/pkg/sql/pg_catalog.go#L409)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: don't panic, instead return errors.AssertionFailedf here
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 355 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-MhnmWUt6S8D28yi2DJp-r18-355:-Mhq6xao5CAqygRHiL_x:b7vgl1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/64cb71f8faa40dde3e0ad8adf0052547f79bcf6f/pkg/sql/logictest/testdata/logic_test/pg_catalog#L355)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm, what if you try without that `TESTCONFIG` variable? (also try with `FILES='pg_catalog|information_schema'` since that file might also need an update)
</blockquote></details>

It works, thanks! Done.
___
*[pkg/sql/opt/optbuilder/testdata/create_table, line 92 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho4AgG20XxftfV4kGV:-Mhq5yYpDn71xYLwTUcR:b9rwaaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/pkg/sql/opt/optbuilder/testdata/create_table#L92)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: is there a way you can remove the spaces so it looks like: `(START 2 INCREMENT 3 CACHE 10)`
</blockquote></details>

We're calling `func (node *SequenceOptions) Format(ctx *FmtCtx)` , which by default add leading and ending spaces to the string of the sequence option. It might be too complicated to change it for now. Added a [TODO](https://github.com/cockroachdb/cockroach/pull/69107/commits/fe03dce0c294d01d34cba029af5c0902bbf5fc18#diff-01f73e0f6ec31b2bbaef6117ca07bc2a1870550909a5ec2bfe8b011387ad4f67R752-R758) for it.
___
*[pkg/sql/parser/sql.y, line 6794 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/69107#-Mho3stg1zZyyv9goI1L:-Mhq6ZpZ15txIU2MKK3_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c37642f68e56790c9cee37b8546f469b94fe213/pkg/sql/parser/sql.y#L6794)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: still a couple tabs
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice this is exciting!
